### PR TITLE
Infra mvp tower

### PR DESCRIPTION
**Pull request by udai-arneja**
# Summary

Simulation of tower without messages complete. Including; tower&agent initiation, agent death and replacement, agent eating, health and floor access, platform moving etc.

## Additional Information

Large change in overall structure, including making a local copy of the abm library, having the loop iterations running in simulation (rather than abm) and the tower holding some core agent information. There is room to streamline and remove redundant structures, but for the mvp this is a stable release.

## Test Plan
Tested with various inputs and varying lengths of time/iterations. Further tests to be written and pushed, but has been tested well for stability.

### CODE REVIEW

**Review by Silvertrousers — APPROVED** [review]
Seems stable, will add message passing later this eve, its likely good enough for mvp just so other teams can get working

**Comment by jzzheng22** [comment]
Why are we mixing between `float64` and `int` for food and HP? I see no reason for food to be a float?

Also, there isn't any reason to specify `64` bits, just use `float`. Also, we should be using `int` for everything anyway, I don't see any reason to use floats or `uint` either.

File: pkg/agents/baseagent.go
Comment (JaafarRammal) [inline pkg/agents/baseagent.go:53]: This is a big issue, overrides can basically be done implicitly on the original `IsDead` function and return wrong states. We should find a way to make these functions final or non-writeable

image

File: pkg/utils/abm/abm.go
Comment (JaafarRammal) [inline pkg/utils/abm/abm.go:115]: If I inject a false ID I can kill a different agent on the next iteration given how the function is written 👀

**Comment by JaafarRammal** [comment]
> Why are we mixing between `float64` and `int` for food and HP? I see no reason for food to be a float?
> 
> Also, there isn't any reason to specify `64` bits, just use `float`. Also, we should be using `int` for everything anyway, I don't see any reason to use floats or `uint` either.

These can be changed later, I am assuming it is temporary design decisions to allow flexibility for agents teams for the current design

File: pkg/infra/tower/tower.go
Comment (jzzheng22) [inline pkg/infra/tower/tower.go:14]: I think if you change this to a map of pointers to structs then you can make `setHP()` a lot cleaner i.e. it would become:
```
tower.agents[id].hp = newHP
```

Comment (jzzheng22) [reply pkg/infra/tower/tower.go:14]: As well as most of your other setter functions in tower-agent.go I think

File: pkg/infra/tower/tower-functions.go
Comment (jzzheng22) [inline pkg/infra/tower/tower-functions.go:50]:
```suggestion
func (t *Tower) reshuffle(numOfFloors int) {
	remainingVacancies := make([]int, numOfFloors)
	log.Printf("Reshuffling alive agents...")

	for i := 0; i < numOfFloors; i++ { // adding a max to each floor
		remainingVacancies[i] = t.agentsPerFloor
	}

	// allocating agents to floors randomly
	// iterate through the uuid strings of each agent
	for id := range t.agents {
		newFloor := rand.Intn(numOfFloors)

		for remainingVacancies[newFloor] == 0 {
			newFloor = rand.Intn(numOfFloors)
		}
		t.mx.RLock()
		currHP := t.agents[id].hp
		aType := t.agents[id].agentType
		t.mx.RUnlock()
		// currHP++
		// aType++
		t.setFloor(id, currHP, newFloor+1, aType)
		remainingVacancies[newFloor]--
	}
}
```

File: pkg/infra/tower/tower-functions.go
Comment (jzzheng22) [inline pkg/infra/tower/tower-functions.go:23]: Why do you need a list of `missingAgents`? Surely an easier way of doing this is just in `simulation.go` `simulationLoop()` you check for which agents have `hp <= 0` and replace those agents??? That way you don't need to keep track of `missingAgents`? This just seems to me like you've got a lot of extra state you have to maintain and keep consistent?

Comment (jzzheng22) [reply pkg/infra/tower/tower-functions.go:23]: This would change how `hpDecay()` is written slightly... but I'm also not convinced `hpDecay()` should be concerned about whether an agent dies or not - it should really just be responsible for decaying HP and nothing else.

File: pkg/infra/tower/tower-functions.go
Comment (jzzheng22) [inline pkg/infra/tower/tower-functions.go:82]:
```suggestion
	if t.agents[id].floor == t.currPlatFloor {
```

File: pkg/utils/abm/abm.go
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:13]:
```suggestion
	currIteration     int
```

File: pkg/utils/abm/abm.go
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:46]:
```suggestion
func (a *ABM) AddAgent(agent Agent) {
	a.mx.Lock()
	defer a.mx.Unlock()
	a.agents = append(a.agents, agent)
}
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]
A lot of changes:

- [x] Fix all of the places I've commented
- [x] Change floats to ints where possible
- [x] If you absolutely insist on using floats, make sure they're of type `float` and not `float64`
- [x] Same goes with `uint`: these should all be `int`. Do not specify their sizes either.
- [x] A lot of abm functions seem to be defined but not used... but I suppose they can stay for now.
- [x] Not sure how many of the mutexes are actually necessary... but I suppose they can also stay for now.

File: pkg/utils/abm/abm.go
Comment (jzzheng22) [reply pkg/utils/abm/abm.go:115]: Irrelevant for MVP I think: focus on good design and working simulation rather than safety/legality of agent actions

File: pkg/agents/baseagent.go
Comment (jzzheng22) [reply pkg/agents/baseagent.go:53]: Irrelevant for MVP I think: focus on good design and working simulation rather than safety/legality of agent actions

**Comment by jzzheng22** [comment]
I feel like a lot of these issues have come from the architectural design - tower shouldn't have this much control over an agent, and agent shouldn't really need to have the tower as part of its struct.

If anything, a lot of these functions arguably belong to the simulation environment which should control the agent and the tower, rather than the tower controlling the agent but then the agent also affecting the tower.

**Comment by JaafarRammal** [comment]
> I feel like a lot of these issues have come from the architectural design - tower shouldn't have this much control over an agent, and agent shouldn't really need to have the tower as part of its struct.
> 
> If anything, a lot of these functions arguably belong to the simulation environment which should control the agent and the tower, rather than the tower controlling the agent but then the agent also affecting the tower.
> 
> I feel like agents should be able to call certain tower functions such as `RequestFood()`, but the setting and decay of HP etc. should be done by the simulation environment, rather than the tower.

I do agree that some architectural decision can be made much better in terms of abstraction and state control. Yet bear in mind none of us has done GoLang before, and more than half the time was spent on understanding how the hell implementing polymorphism, virtual functions, overrides, finals, or just inheritance would work safely. All of that, and with a deadline in mind to deliver as quickly as possible to other teams something that works so they can start experimenting on their agents.

The current PR is **an MVP** and not a final deliverable. This means it is subject to changes (some can be major), but at least it works and unblocks the agent teams to create simulations and start experimenting. All the architectural **changes can be done without distributing the agent teams** given the current abstraction (they have a run function and a few getters which can be modified from base if we need to make changes)

> a lot of these functions arguably belong to the simulation environment

Yes and they can be moved later, it won't disrupt the agent teams work

> setting and decay of HP etc. should be done by the simulation environment, rather than the tower

Yes and that can be done later

File: pkg/agents/baseagent.go
Comment (JaafarRammal) [reply pkg/agents/baseagent.go:53]: It will make it harder to review code since we have to check people don't override by mistake. Will leave it away for now though since we need to get a working MVP out

File: pkg/utils/abm/abm.go
Comment (JaafarRammal) [reply pkg/utils/abm/abm.go:115]: Nothing to do with safety / legality, it's a bug in the function

File: pkg/infra/simulation/simulation.go
Comment (JaafarRammal) [reply pkg/infra/simulation/simulation.go:18]: Only main accesses simEnv

File: pkg/agents/baseagent.go
Comment (JaafarRammal) [reply pkg/agents/baseagent.go:54]: Added TODO

File: pkg/utils/abm/abm.go
Comment (JaafarRammal) [reply pkg/utils/abm/abm.go:116]: Added TODO

File: pkg/infra/tower/tower-functions.go
Comment (JaafarRammal) [reply pkg/infra/tower/tower-functions.go:23]: 1- Agreed. Future version we shall move states to simulation
2- Eventually `hpDecay()` would be parametrized, we'd like agent teams to PR proposals for it. We check in the function for death directly as otherwise we'd be checking right after it in a different loop without any agent state. Check can be moved to a different loop without issues

File: pkg/infra/simulation/simulation.go
Comment (JaafarRammal) [reply pkg/infra/simulation/simulation.go:93]: probably random HP do experiment on agents dying and testing, all agents should have same starting HP. replaced

File: pkg/infra/tower/tower.go
Comment (JaafarRammal) [reply pkg/infra/tower/tower.go:50]: can be modified as wished in another PR

File: pkg/infra/tower/tower.go
Comment (JaafarRammal) [reply pkg/infra/tower/tower.go:14]: TODOed

**Review by JaafarRammal — APPROVED** [review]
Changes can be made to architecture, but it works and unblocks everyone and has all features required for MVP setup, so LGTM

**Comment by jzzheng22** [comment]
Thanks for all the changes, they look good. Holding off on approval until we look at https://github.com/SOMAS2021/SOMAS2021/tree/infra-mvp-rework and see if message passing can be integrated into that quickly.

**Review by JaafarRammal — APPROVED** [review]

**Review by jzzheng22 — APPROVED** [review]
LGTM, thanks all.

**Comment by JaafarRammal** [comment]
> Thanks for all the changes, they look good. Holding off on approval until we look at https://github.com/SOMAS2021/SOMAS2021/tree/infra-mvp-rework and see if message passing can be integrated into that quickly.

Happy to have a look at that branch as refactoring separately from this PR as it skips on some design decisions and access control

**Pull request merged**
